plugins/clickhouse/src/transformers.cpp

```diff
@@ -0,0 +1,644 @@
+    }                                                                          \
+  }
+
+X(int64_type, ColumnInt64, "Int64");
+X(uint64_type, ColumnUInt64, "UInt64");
+X(double_type, ColumnFloat64, "Float64");
+X(string_type, ColumnString, "String");
+X(duration_type, ColumnInt64, "Int64");
+X(ip_type, ColumnIPv6, "IPv6");
+#undef X
+
+template <>
+struct tenzir_to_clickhouse_trait<time_type> {
+  constexpr static std::string_view name = "DateTime64(9)";
+  using column_type = ColumnDateTime64;
+
+  constexpr static auto null_value = std::nullopt;
+
+  static auto clickhouse_typename(bool nullable) -> std::string {
+    if (nullable) {
+      return std::string{"Nullable("}.append(name).append(")");
+    }
+    return std::string{name};
+  }
+
+  template <bool nullable>
+  static auto allocate(size_t n) {
+    using Column_Type
+      = std::conditional_t<nullable, ColumnNullableT<column_type>, column_type>;
+    auto res = std::make_shared<Column_Type>(9);
+    res->Reserve(n);
+    return res;
+  }
+};
+
+template <>
+struct tenzir_to_clickhouse_trait<subnet_type> {
+  static auto clickhouse_typename(bool nullable) -> std::string {
+    if (nullable) {
+      return "Tuple(ip Nullable(IPv6),length Nullable(UInt8))";
+    }
+    return "Tuple(ip IPv6,length UInt8)";
+  }
+
+  constexpr static auto null_value = std::tuple{std::nullopt, std::nullopt};
+
+  template <bool nullable>
+  static auto allocate(size_t n) {
+    using ip_t
+      = std::conditional_t<nullable, ColumnNullableT<ColumnIPv6>, ColumnIPv6>;
+    using length_t
+      = std::conditional_t<nullable, ColumnNullableT<ColumnUInt8>, ColumnUInt8>;
+    using Column_Type = ColumnTupleT<ip_t, length_t>;
+    auto ip_c = std::make_shared<ip_t>();
+    ip_c->Reserve(n);
+    auto length_c = std::make_shared<length_t>();
+    length_c->Reserve(n);
+    auto res = std::make_shared<Column_Type>(
+      std::tuple{std::move(ip_c), std::move(length_c)});
+    return res;
+  }
+};
+
+template <typename Expected, typename Actual>
+concept convertible_hack = std::same_as<Expected, Actual>
+                           or (std::same_as<Expected, int64_type>
+                               and std::same_as<Actual, duration_type>);
+
+template <typename T, bool Nullable>
+  requires requires { tenzir_to_clickhouse_trait<T>{}; }
+struct transformer_from_trait : transformer {
+  using traits = tenzir_to_clickhouse_trait<T>;
+
+  transformer_from_trait()
+    : transformer{traits::clickhouse_typename(Nullable), Nullable} {
+  }
+
+  virtual auto update_dropmask(const tenzir::type&, const arrow::Array& array,
+                               dropmask_ref dropmask,
+                               tenzir::diagnostic_handler&) -> drop override {
+    if constexpr (Nullable) {
+      return drop::none;
+    }
+    if (not array.null_bitmap()) {
+      return drop::none;
+    }
+    if (array.null_count() == 0) {
+      return drop::none;
+    }
+    // TODO we can potentially discover a "drop::all" here
```

Comment:
```suggestion
    if (array.null_count() == array.length()) { … }
```

This same TODO also exists in a few other places in this file.